### ISSUE TITLE
[C] Segmentation fault when reading meta info

### ISSUE THREAD

**Issue by chovyy** [eth-cscs/serialbox2#131]
In the Fortran unit test everything works fine when it calls `fs_get_halos`, but in a real world program I'm receiving a segmentation fault from [`serialbox-c/FieldMetainfo.cpp`](https://github.com/eth-cscs/serialbox2/blob/master/src/serialbox-c/FieldMetainfo.cpp).

```
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x7fe89cc8216a in ???
#1  0x7fe89cc81393 in ???
#2  0x7fe89c69f13f in ???
#3  0x7fe89d1fb1d2 in ???
#4  0x7fe89d1fabb3 in serialboxFieldMetainfoGetMetainfo
        at /home/christian/workspace/serialbox2/src/serialbox-c/FieldMetainfo.cpp:104
#5  0x7fe89d20c28b in serialboxSerializerGetFieldMetainfo2
        at /home/christian/workspace/serialbox2/src/serialbox-c/Serializer.cpp:337
#6  0x7fe89d22128a in serialboxFortranSerializerGetFieldHalos
        at /home/christian/workspace/serialbox2/src/serialbox-c/FortranWrapper.cpp:194
#7  0x7fe89d1e8862 in __m_serialize_MOD_fs_get_halos
        at /home/christian/workspace/serialbox2/src/serialbox-fortran/m_serialize.f90:928
#8  0x558cc6f55501 in test
        at /home/christian/workspace/serialbox2-metainfo-bug/test.f90:46
#9  0x558cc6f55657 in main
        at /home/christian/workspace/serialbox2-metainfo-bug/test.f90:3
```
I have created an example including the files from the unit test that work fine and the one that causes the crash. I kept the example in Fortran because it's just easier to get a stack trace there, but I have tried to use the C library directly and it also ended up in a seg fault.

[serialbox2-metainfo-bug.zip](https://github.com/eth-cscs/serialbox2/files/1605992/serialbox2-metainfo-bug.zip)

**Comment by havogt** [comment]
Sorry, I didn't have time to look into it yet. Is it blocking you?

**Comment by chovyy** [comment]
Yes, but I have enough other things to do, so no worries.

**Comment by vectorflux** [comment]
I know Hannes is really busy on the COSMO-GT work.  It would be really helpful if we could move FTG to Serialbox2, since -- as far as I know -- it is the only software which still relies on the old Serialbox (which is to be decommissioned).  A diploma student starts working with FTG this week, if possible using the Serialbox2 version

**Comment by havogt** [comment]
Oh shit. I missed your reply @chovyy. I will have a look today. Maybe it is a quick fix.

**Comment by vectorflux** [comment]
No big deal.  It would just be nice to decommission Serialbox...

**Comment by havogt** [comment]
I cannot reproduce the segfault. I just did `make` and got the following output:
```
 double0
           0           0           0           0           0           0           0           0
 ================================================================================================
 double1
          -1           6           0           0           0           0           0           0
 ================================================================================================
 double2
          -2           2          -3           8           0           0           0           0
 ================================================================================================
 double3
          -4           4          -5           5          -6          11           0           0
 ================================================================================================
 double4
           7           8          23          42          12          -4           0          21
 ================================================================================================
 p_cc
           1           8           1          47           1        1330           0           0
 ================================================================================================
```

**Comment by chovyy** [comment]
Which compiler, OS?

**Comment by havogt** [comment]
On Linux with gcc 7.2.1 (g++ and gfortran).

**Comment by havogt** [comment]
Do you use the most recent serialbox2 version? Do you use the `SERIALBOX_ENABLE_EXPERIMENTAL_FILESYSTEM=ON` (which should be the default).

**Comment by chovyy** [comment]
Yes, I do. I'm working with `gcc (Ubuntu 7.2.0-8ubuntu3) 7.2.0`, but have also tried older versions. Very strange...

**Comment by havogt** [comment]
looks very similar to my setup

**Comment by havogt** [comment]
Maybe someone else can try to reproduce? @vectorflux ;)

**Comment by chovyy** [comment]
My apologies, it seems to be completely my fault.

1st. The error was actually only in the Fortran interface. I don't know, how I came to the conclusion that it also occurred when I was using the C interface.

2nd. I have observed a similar error before, but forgot to report it.

3rd. I have now found out that I introduced that error myself by missing `//C_NULL_CHAR` when passing the fieldname to the wrapper routines within `fs_get_size` and `fs_get_halos`.

I have created a [pull request](https://github.com/eth-cscs/serialbox2/pull/133) that should fix that issue.
